stablehlo/transforms/ChloDecompositionPatternsMath.td

```diff
@@ -384,3 +384,544 @@ def : Pat<(CHLO_AsinOp NonComplexElementType:$x),
+                          (STABLEHLO_DEFAULT_COMPARISON_TYPE)),
+                        (StableHLO_MulOp $sqrt_two, $_hypot_2_mx),
+                        (StableHLO_SelectOp
+                          (StableHLO_AndOp
+                            (StableHLO_CompareOp
+                             (StableHLO_SqrtOp:$_hypot_2_sqa
+                               (StableHLO_AddOp
+                                 $one,
+                                 (StableHLO_MulOp:$_hypot_2_r
+                                   (StableHLO_DivOp:$_hypot_2_mn_over_mx $_hypot_2_mn, $_hypot_2_mx),
+                                   $_hypot_2_mn_over_mx))),
+                             $one,
+                              StableHLO_ComparisonDirectionValue<"EQ">,
+                              (STABLEHLO_DEFAULT_COMPARISON_TYPE)),
+                            (StableHLO_CompareOp
+                             $_hypot_2_r,
+                             $zero,
+                              StableHLO_ComparisonDirectionValue<"GT">,
+                              (STABLEHLO_DEFAULT_COMPARISON_TYPE))),
+                          (StableHLO_AddOp
+                            $_hypot_2_mx,
+                            (StableHLO_DivOp
+                              (StableHLO_MulOp $_hypot_2_mx, $_hypot_2_r),
+                              $two)),
+                          (StableHLO_MulOp $_hypot_2_mx, $_hypot_2_sqa))))),
+                  $x)),
+              (StableHLO_AddOp
+                (StableHLO_DivOp
+                  (StableHLO_MulOp:$yy $y, $y),
+                  (StableHLO_AddOp:$rpxp1 $r, $xp1)),
+                (StableHLO_SubtractOp:$smxm1 $s, $xm1)))),
+          (StableHLO_MulOp
+            $y,
+            (StableHLO_SqrtOp
+              (StableHLO_AddOp
+                (StableHLO_DivOp $half_apx, $rpxp1),
+                (StableHLO_DivOp
+                  $half_apx,
+                  (StableHLO_AddOp:$spxm1 $s, $xm1))))))),
+      $signed_x),
+    (StableHLO_SelectOp
+      (StableHLO_CompareOp
+       $signed_y,
+       $zero,
+        StableHLO_ComparisonDirectionValue<"LT">,
+        (STABLEHLO_DEFAULT_COMPARISON_TYPE)),
+      (StableHLO_SelectOp:$imag
+        (StableHLO_CompareOp
+         (StableHLO_SelectOp:$mx
+           (StableHLO_CompareOp:$y_gt_safe_max_opt
+            $y,
+            (StableHLO_SelectOp:$safe_max_opt
+              (StableHLO_CompareOp
+               $x,
+               (StableHLO_MulOp
+                 $safe_max,
+                 (StableHLO_ConstantLike<"1000000000000.0"> $signed_x)),
+                StableHLO_ComparisonDirectionValue<"LT">,
+                (STABLEHLO_DEFAULT_COMPARISON_TYPE)),
+              (StableHLO_MulOp
+                $safe_max,
+                (StableHLO_ConstantLike<"1e-06"> $signed_x)),
+              (StableHLO_MulOp
+                $safe_max,
+                (StableHLO_ConstantLike<"100.0"> $signed_x))),
```

Comment:
Very difficult to understand what's going on in this decomposition, especially with these large/small constants which don't appear in the above formula. I'm not sure how these behave with different (smaller) datatypes?

Any way to simplify this to read more like the math? Can we generate other CHLO ops in the decomposition? I.e. I believe we can generate `CHLO_Acos(x/a)` as a part of this instead of decomposing inline (assuming that's the real portion which would then be decomposed by the following decomp).

Reply:
> Very difficult to understand what's going on in this decomposition, especially with these large/small constants which don't appear in the above formula.

These constants are defined in:
https://github.com/pearu/stablehlo/blob/7c655d3b5df4c6a641590299f7626a6c9e8a5fcd/stablehlo/transforms/ChloDecompositionPatternsMath.td#L124
(`100.0` is `1e2`).

> I'm not sure how these behave with different (smaller) datatypes?

This is a good question. The validity of constant `safe_max_opt` is currently tested against float32 and float64 data types. Since the definition of `safe_max_opt` worked fine for float32 and float64, I'd expect that the used scaling works fine also for `float16`. However, I agree that the validity of `safe_max_opt` needs to be tested against float16 as well. I propose to address this in a follow-up PR.

For bfloat16, we don't have tests because numpy does not support such dtype. That said, the tests for bfloat16 could be still generated, this just needs to be addressed within functional_algorithms (issue https://github.com/pearu/functional_algorithms/issues/13) by avoiding the convenience of numpy availability...

> Any way to simplify this to read more like the math? Can we generate other CHLO ops in the decomposition? I.e. I believe we can generate CHLO_Acos(x/a) as a part of this instead of decomposing inline (assuming that's the real portion which would then be decomposed by the following decomp).

Yes, there is a way to simplify the currently generated code because the algorithm for all the four complex functions - asin, asinh, acos, acosh - use the same kernel algorithm (see below).

I agree that reading this generated code for math is difficult. Instead, I'd suggest reading the original definitions of these algorithms which expose the mathematical definitions of the algorithms in much more clearly. For instance,
- https://github.com/pearu/functional_algorithms/blob/df060980aa5347bb695a66be44232a6bbae91bfe/functional_algorithms/algorithms.py#L257 defines the kernel algorithm `complex_asin_acos(z: complex)` mentioned above
- `asin(z: complex)` is evaluated as `complex_asin_acos(z)[0]`
- `acos(z: complex)` is evaluated as `complex_asin_acos(z)[1]`
- `asinh(z: complex)` is evaluated as `-I * asin(I * z)` (see https://github.com/pearu/functional_algorithms/blob/df060980aa5347bb695a66be44232a6bbae91bfe/functional_algorithms/algorithms.py#L441)
- `acosh(z: complex)` is evaluated as `sqrt(z - 1) / sqrt(1 - z) * acos(z)` (see https://github.com/pearu/functional_algorithms/blob/df060980aa5347bb695a66be44232a6bbae91bfe/functional_algorithms/algorithms.py#L517)

It would be possible to refactor the code generation for StableHLO so that the kernel algorithm would be exposed as the only "difficult-to-read" decomposition and the decompositions of asin, acos, asinh, acosh would call the kernel operation as defined in the above in FA. Could we address this as a follow-up task as it would require refactoring the currently landed implementation of `asin` as well?

What do you think?

Reply:
Meanwhile, I have implemented `asin_acos_kernel` in functional_algorithms (https://github.com/pearu/functional_algorithms/pull/14) that allows representing all functions - asin, acos, asinh, acosh - in terms of a single kernel function. It will be available in the next release of functional_algorithms.

Reply:
> Yes, there is a way to simplify the currently generated code because the algorithm for all the four complex functions - asin, asinh, acos, acosh - use the same kernel algorithm (see below).

This would be a great refactor. The more re-use in decompositions the better, if we can generate other CHLO ops used in the intermediate phases would be great, if we want to add a CHLO_Hypot (if that would make codegen smaller/more readable) we can, etc. Just thinking about general maintenance of these patterns.

In the meantime yes, we can make this a followup.

Reply:
Resolving via followup task https://github.com/pearu/functional_algorithms/issues/15